frame/assets/src/lib.rs

```diff
@@ -1088,546 +1340,123 @@ impl<T: Config> Pallet<T> {
 
-	#[test]
-	fn transferring_frozen_user_should_not_work() {
-		new_test_ext().execute_with(|| {
-			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
-			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
-			assert_eq!(Assets::balance(0, 1), 100);
-			assert_ok!(Assets::freeze(Origin::signed(1), 0, 1));
-			assert_noop!(Assets::transfer(Origin::signed(1), 0, 2, 50), Error::<Test>::Frozen);
-			assert_ok!(Assets::thaw(Origin::signed(1), 0, 1));
-			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 50));
-		});
-	}
+			let metadata = Metadata::<T>::take(&id);
+			T::Currency::unreserve(&details.owner, details.deposit.saturating_add(metadata.deposit));
 
-	#[test]
-	fn transferring_frozen_asset_should_not_work() {
-		new_test_ext().execute_with(|| {
-			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
-			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
-			assert_eq!(Assets::balance(0, 1), 100);
-			assert_ok!(Assets::freeze_asset(Origin::signed(1), 0));
-			assert_noop!(Assets::transfer(Origin::signed(1), 0, 2, 50), Error::<Test>::Frozen);
-			assert_ok!(Assets::thaw_asset(Origin::signed(1), 0));
-			assert_ok!(Assets::transfer(Origin::signed(1), 0, 2, 50));
-		});
-	}
+			Approvals::<T>::remove_prefix(&id);
+			Self::deposit_event(Event::Destroyed(id));
 
-	#[test]
-	fn origin_guards_should_work() {
-		new_test_ext().execute_with(|| {
-			assert_ok!(Assets::force_create(Origin::root(), 0, 1, 10, 1));
-			assert_ok!(Assets::mint(Origin::signed(1), 0, 1, 100));
-			assert_noop!(Assets::transfer_ownership(Origin::signed(2), 0, 2), Error::<Test>::NoPermission);
-			assert_noop!(Assets::set_team(Origin::signed(2), 0, 2, 2, 2), Error::<Test>::NoPermission);
-			assert_noop!(Assets::freeze(Origin::signed(2), 0, 1), Error::<Test>::NoPermission);
-			assert_noop!(Assets::thaw(Origin::signed(2), 0, 2), Error::<Test>::NoPermission);
-			assert_noop!(Assets::mint(Origin::signed(2), 0, 2, 100), Error::<Test>::NoPermission);
-			assert_noop!(Assets::burn(Origin::signed(2), 0, 1, 100), Error::<Test>::NoPermission);
-			assert_noop!(Assets::force_transfer(Origin::signed(2), 0, 1, 2, 100), Error::<Test>::NoPermission);
-			assert_noop!(Assets::set_max_zombies(Origin::signed(2), 0, 11), Error::<Test>::NoPermission);
-			assert_noop!(Assets::destroy(Origin::signed(2), 0, 100), Error::<Test>::NoPermission);
-		});
+			// NOTE: could use postinfo to reflect the actual number of accounts/sufficient/approvals
+			Ok(())
+		})
 	}
 
-	#[test]
-	fn transfer_owner_should_work() {
-		new_test_ext().execute_with(|| {
-			Balances::make_free_balance_be(&1, 100);
-			Balances::make_free_balance_be(&2, 1);
-			assert_ok!(Assets::create(Origin::signed(1), 0, 1, 10, 1));
+	fn do_transfer(
+		id: T::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: T::Balance,
+		maybe_need_admin: Option<T::AccountId>,
+	) -> DispatchResult {
+		ensure!(!amount.is_zero(), Error::<T>::AmountZero);
```

Comment:
If we want to follow more closely to ERC20 specification, zero transfer transactions should execute successfully and emit a transferred event